### ISSUE TITLE
Generating Javadoc in a Maven project using <detectLinks> does not work

### ISSUE THREAD

**Issue by alerosmile** [assertj/assertj#3476]
The `maven-javadoc-plugin` has an option [detectLinks](https://maven.apache.org/plugins/maven-javadoc-plugin/jar-mojo.html#detectLinks) which enables detection of Javadoc links of dependent projects. The detection works only if the dependent project follows the default Maven conventions (javadoc must be located under `${project.url}/apidocs`).

The project URL of assertj-core is `https://assertj.github.io/doc/#assertj-core`. The Maven plugin first checks if the URL `https://assertj.github.io/doc/#assertj-core/apidocs/element-list` is accessible and then forwards the URL `'https://assertj.github.io/doc/#assertj-core/apidocs'` to the javadoc tool. 

But the javadoc tool interprets the `#` character, reads the URL `https://assertj.github.io/doc/package-list` and obviously fails:
```
Error fetching URL: https://assertj.github.io/doc/#assertj-core/apidocs/ (java.io.FileNotFoundException: https://assertj.github.io/doc/package-list)
```

I can workaround this by defining in my project:
```xml
  <plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-javadoc-plugin</artifactId>
    <configuration>
      <dependencyLinks>
        <dependencyLink>
          <groupId>org.assertj</groupId>
          <artifactId>assertj-core</artifactId>
          <url>https://www.javadoc.io/doc/org.assertj/assertj-core/${assertj.version}</url>
        </dependencyLink>
      </dependencyLinks>
    </configuration>
  </plugin>
```

Maybe the project URL of assertj-core should be `https://www.javadoc.io/doc/org.assertj/assertj-core/${assertj.version}` instead of `https://assertj.github.io/doc/#assertj-core`. But then the javadoc must be located under `https://www.javadoc.io/doc/org.assertj/assertj-core/${assertj.version}/apidocs`.

**Comment by scordio** [comment]
> Maybe the project URL of assertj-core should be `https://www.javadoc.io/doc/org.assertj/assertj-core/${assertj.version}` instead of `https://assertj.github.io/doc/#assertj-core`. But then the javadoc must be located under `https://www.javadoc.io/doc/org.assertj/assertj-core/${assertj.version}/apidocs`.

I wouldn't point the project URL to `javadoc.io`. However, I see the problem.

We might deploy the Javadoc following the `{project.url}/apidocs` convention, where `project.url` would still belong to `assertj.github.io`.

However, I checked how some other projects behave and none of them seems to do it according to the Maven conventions:
* JUnit: [expected](https://junit.org/junit5/apidocs) vs. [actual](https://junit.org/junit5/docs/current/api/)
* Mockito: [expected](https://github.com/mockito/mockito/apidocs) vs. [actual](https://javadoc.io/doc/org.mockito/mockito-core/latest/)
* Spring Framework: [expected](https://github.com/spring-projects/spring-framework/apidocs) vs. [actual](https://docs.spring.io/spring-framework/docs/current/javadoc-api/)

**Comment by alerosmile** [comment]
@scordio You are right. Many other projects do not follow the convention, but none of them has broken my build so far.

In other projects, the Maven plugin detects the 'invalid' links and doesn't forward them to the javadoc tool. Maybe the Maven plugin should interpret the `#` character the same way the javadoc plugin does.

Currently, my workaround seems to be the way to go.

**Comment by scordio** [comment]
We will rework our docs to support v3 and v4 in parallel (and maybe relocate them somewhere else in our GitHub org) so I'm leaving this open for further thoughts.

**Comment by scordio** [comment]
Just FYI, we use a slightly shorter config in `assertj-guava`:

https://github.com/assertj/assertj/blob/4f072c99641691f67f4217eff44d4d1866ddc274/assertj-guava/pom.xml#L168-L171

And I just realized we have a cosmetic bug 🙂 (#3478)

**Comment by alerosmile** [comment]
You have to use [`<dependencyLinks>`](https://maven.apache.org/plugins/maven-javadoc-plugin/jar-mojo.html#dependencyLinks) if [`<detectLinks>`](https://maven.apache.org/plugins/maven-javadoc-plugin/jar-mojo.html#detectLinks) is enabled.

The [`<links>`](https://maven.apache.org/plugins/maven-javadoc-plugin/jar-mojo.html#links) option just adds additional links.

**Comment by michael-o** [comment]
It is not clear to me the expectation is. Both `https://assertj.github.io/doc/#assertj-core/apidocs/package-list` and `https://assertj.github.io/doc/package-list` point to non-existing resources. Please explain?

**Comment by alerosmile** [comment]
@michael-o Not sure what you mean. For me, both URLs are accessible. But the URL including the `#` are is interpreted differently by the javadoc tool (see my first post).

**Comment by michael-o** [comment]
> @michael-o Not sure what you mean. For me, both URLs are accessible. But the URL including the `#` are is interpreted differently by the javadoc tool (see my first post).

I need to correct myself: Both resources point to existing, but invalid resources:
```
mosipov@mikaw10 MINGW64 ~
$ curl 'https://assertj.github.io/doc/#assertj-core/apidocs/package-list' | head
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  1  926k    1 16375    0     0   227k      0  0:00:04 --:--:--  0:00:04  228k<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.10">
<link rel="icon" type="image/png" href="./images/favicon.png">
<title>AssertJ - fluent assertions java library</title>
<style>
  3  926k    3 32750    0     0   428k      0  0:00:02 --:--:--  0:00:02  432k
curl: (23) Failure writing output to destination

mosipov@mikaw10 MINGW64 ~
$ curl 'https://assertj.github.io/doc/package-list' | head
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-type" content="text/html; charset=utf-8">
    <meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src 'unsafe-inline'; img-src data:; connect-src 'self'">
    <title>Page not found &middot; GitHub Pages</title>
    <style type="text/css" media="screen">
      body {
        background-color: #f1f1f1;
        margin: 0;
 97  9379   97  9173    0     0  62047      0 --:--:-- --:--:-- --:--:-- 61979
curl: (23) Failure writing output to destination
```
Neither a fixed Maven Javadoc Plugin nor Javadoc itself can use it...

**Comment by alerosmile** [comment]
Whether the resources are valid or not is not the point in this case.

It is important that the Maven plugin and the Javadoc tool interpret the URL identically. The Maven plugin checks whether the URL is accessible. If yes, the URL is added to the options file for the Javadoc tool.

In this case the Maven plugin checks the URL `https://assertj.github.io/doc/#assertj-core/apidocs/element-list`. It is accessible and therefore added to the options file. But the Javadoc tool does not access the same URL, it tries to read `https://assertj.github.io/doc/package-list` which is not accessible and therefore the Javadoc tool fails.

In order to not fail the build, only URLs which are accessible by the Javadoc tool should be added to the options file. For this to work properly, the Maven plugin must interpret the URLs exactly the same way the Javadoc tool does.

**Comment by michael-o** [comment]
> Whether the resources are valid or not is not the point in this case.
> 
> It is important that the Maven plugin and the Javadoc tool interpret the URL identically. The Maven plugin checks whether the URL is accessible. If yes, the URL is added to the options file for the Javadoc tool.
> 
> In this case the Maven plugin checks the URL `https://assertj.github.io/doc/#assertj-core/apidocs/element-list`. It is accessible and therefore added to the options file. But the Javadoc tool does not access the same URL, it tries to read `https://assertj.github.io/doc/package-list` which is not accessible and therefore the Javadoc tool fails.
> 
> In order to not fail the build, only URLs which are accessible by the Javadoc tool should be added to the options file. For this to work properly, the Maven plugin must interpret the URLs exactly the same way the Javadoc tool does.

Agreed, or the plugin should not check the resource at all and leave everything to Javadoc. At the end, no one needs duplicate checks w/o any benefits. PRs welcome.

**Comment by michael-o** [comment]
What I can offer is to broaden the validation scope of the `validateLinks` paramater not to touch URLs at all, instead of getting them and not validating content... maybe that would be most reasonable. Why fetch if you don't check the content. It can be deceiving.

**Comment by michael-o** [comment]
Prepare a test branch for you guys...

**Comment by michael-o** [comment]
Please try a custom build from this PR: https://github.com/apache/maven-javadoc-plugin/pull/308

**Comment by michael-o** [comment]
Darn, while working on the PR I have found a real bug in a PR which was not visible before...

**Comment by alerosmile** [comment]
Just looking at your PR raises the same question again (for me):
Assuming the URL `https://assertj.github.io/doc/#assertj-core/apidocs/package-list` and its content are perfectly valid, the Maven plugin forwards the URL to the javadoc tool (validation succeeds). But the javadoc tool does not read the same URL. It reads `https://assertj.github.io/doc/package-list` and fails.

**Comment by michael-o** [comment]
> Just looking at your PR raises the same question again (for me): Assuming the URL `https://assertj.github.io/doc/#assertj-core/apidocs/package-list` and its content are perfectly valid, the Maven plugin forwards the URL to the javadoc tool (validation succeeds). But the javadoc tool does not read the same URL. It reads `https://assertj.github.io/doc/package-list` and fails.

While this is true, at least the Maven Javadoc Plugin does not touch the URL in this regard. We cannot change how `javadoc(1)` operates and I consider the behavior correct. You are manipulating the fragment identifier of the URI, not the URI itself. This will not work for anything non-browser.

**Comment by alerosmile** [comment]
But then the Maven plugin should consider all URLs containing a fragment identifier as invalid an not forward them to the javadoc tool. That would solve this issue.

**Comment by michael-o** [comment]
> But then the Maven plugin should consider all URLs containing a fragment identifier as invalid an not forward them to the javadoc tool. That would solve this issue.

How is the plugin supposed to do that? How to semantically decide that? At best, we parse the URL, chop of the fragment identifier and add `package-list`? Is that what you expect?

**Comment by alerosmile** [comment]
> > But then the Maven plugin should consider all URLs containing a fragment identifier as invalid an not forward them to the javadoc tool. That would solve this issue.
> 
> How is the plugin supposed to do that? How to semantically decide that? At best, we parse the URL, chop of the fragment identifier and add `package-list`? Is that what you expect?

Isn't this exactly what the javadoc tool does? I tried to lookup it's source code but didn't find it.

**Comment by michael-o** [comment]
> > > But then the Maven plugin should consider all URLs containing a fragment identifier as invalid an not forward them to the javadoc tool. That would solve this issue.
> > 
> > 
> > How is the plugin supposed to do that? How to semantically decide that? At best, we parse the URL, chop of the fragment identifier and add `package-list`? Is that what you expect?
> 
> Isn't this exactly what the javadoc tool does? I tried to lookup it's source code but didn't find it.

The source is available, but this isn't what you expect. All it can do it is chop of the FI and do a `package-list` fetch. Whether the actual URL with the FI is semantically correct is undefined, but you expect `https://assertj.github.io/doc/#assertj-core/apidocs/package-list` to be valid, that will never be the case for as long as `package-list` is part of FI.

**Comment by alerosmile** [comment]
The javadoc tool resolves the URL [here](https://github.com/openjdk/jdk/blob/e3a5e265a7747b02b8f828fbedea0dda7246fc51/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Extern.java#L438).
If `elemlisturlpath` is `https://assertj.github.io/doc/#assertj-core/apidocs/` then `link` will be `https://assertj.github.io/doc/element-list`.


And this is what I believe should be done the same way [here ](https://github.com/apache/maven-javadoc-plugin/blob/cb10f7861126983f4cff840f9f8df5e924a66f92/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java#L5788).

`isValidJavadocLink()` checks `https://assertj.github.io/doc/#assertj-core/apidocs/element-list` but should check `https://assertj.github.io/doc/element-list`.

Maybe something like:
`elementListUri = new URI(link).resolve(ELEMENT_LIST);`

**Comment by michael-o** [comment]
> The javadoc tool resolves the URL [here](https://github.com/openjdk/jdk/blob/e3a5e265a7747b02b8f828fbedea0dda7246fc51/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Extern.java#L438). If `elemlisturlpath` is `https://assertj.github.io/doc/#assertj-core/apidocs/` then `link` will be `https://assertj.github.io/doc/element-list`.
> 
> And this is what I believe should be done the same way [here ](https://github.com/apache/maven-javadoc-plugin/blob/cb10f7861126983f4cff840f9f8df5e924a66f92/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java#L5788).
> 
> `isValidJavadocLink()` checks `https://assertj.github.io/doc/#assertj-core/apidocs/element-list` but should check `https://assertj.github.io/doc/element-list`.
> 
> Maybe something like: `elementListUri = new URI(link).resolve(ELEMENT_LIST);`

This makes sense.

**Comment by michael-o** [comment]
Here is an incomplete patch:
```diff
diff --git a/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java b/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
index 0a291dd9..595f81dd 100644
--- a/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
+++ b/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
@@ -5765,6 +5765,9 @@ public abstract class AbstractJavadocMojo extends AbstractMojo {
             try {
                 redirectLinks.add(JavadocUtil.getRedirectUrl(new URI(link).toURL(), settings)
                         .toString());
+            } catch (URISyntaxException | MalformedURLException | IllegalArgumentException e) {
+                // only print in debug, it should have been logged already in warn/error because link isn't valid
+                getLog().debug("Could not follow " + link + ". Reason: " + e.getMessage());
             } catch (IOException e) {
                 // only print in debug, it should have been logged already in warn/error because link isn't valid
                 getLog().debug("Could not follow " + link + ". Reason: " + e.getMessage());
@@ -5773,9 +5776,6 @@ public abstract class AbstractJavadocMojo extends AbstractMojo {
                 // incomplete redirect configuration on the server side.
                 // This partially restores the previous behaviour before fix for MJAVADOC-427
                 redirectLinks.add(link);
-            } catch (URISyntaxException | IllegalArgumentException e) {
-                // only print in debug, it should have been logged already in warn/error because link isn't valid
-                getLog().debug("Could not follow " + link + ". Reason: " + e.getMessage());
             }
         }
         return redirectLinks;
@@ -5799,8 +5799,8 @@ public abstract class AbstractJavadocMojo extends AbstractMojo {
                     || link.trim().toLowerCase(Locale.ENGLISH).startsWith("https:")
                     || link.trim().toLowerCase(Locale.ENGLISH).startsWith("ftp:")
                     || link.trim().toLowerCase(Locale.ENGLISH).startsWith("file:")) {
-                packageListUri = new URI(link + '/' + PACKAGE_LIST);
-                elementListUri = new URI(link + '/' + ELEMENT_LIST);
+                packageListUri = new URI(link).resolve(PACKAGE_LIST);
+                elementListUri = new URI(link).resolve(ELEMENT_LIST);
             } else {
                 // links can be relative paths or files
                 File dir = new File(link);
@@ -5908,9 +5908,11 @@ public abstract class AbstractJavadocMojo extends AbstractMojo {
      * @param p not null
      * @return the javadoc link based on the project url i.e. <code>${project.url}/${destDir}</code> where
      *         <code>destDir</code> is configued in the Javadoc plugin configuration (<code>apidocs</code> by default).
+     * @throws URISyntaxException
+     * @throws MalformedURLException
      * @since 2.6
      */
-    private static String getJavadocLink(MavenProject p) {
+    private static String getJavadocLink(MavenProject p) throws MalformedURLException, URISyntaxException {
         if (p.getUrl() == null) {
             return null;
         }
@@ -5924,7 +5926,7 @@ public abstract class AbstractJavadocMojo extends AbstractMojo {
             destDir = destDirConfigured;
         }

-        return url + "/" + destDir;
+        return new URI(url).resolve(destDir).toURL().toString();
     }
``´
Does not compile ATM.

**Comment by michael-o** [comment]
Scratch that patch, it does not work. The approch taken by the plugin and Javadoc a diametral. This needs to be aligned first.

**Comment by alerosmile** [comment]
@michael-o A long time ago I've created an issue in the Maven Javadoc plugin project:
https://issues.apache.org/jira/projects/MJAVADOC/issues/MJAVADOC-792

Maybe we should continue the discussion there.

**Comment by michael-o** [comment]
> @michael-o A long time ago I've created an issue in the Maven Javadoc plugin project: https://issues.apache.org/jira/projects/MJAVADOC/issues/MJAVADOC-792
> 
> Maybe we should continue the discussion there.

Agree.

**Comment by scordio** [comment]
Assuming we would publish the Javadoc following `detectLinks` expectations, version pinpointing wouldn't be possible without a more complex setup.

See also the conversation with JUnit at https://github.com/junit-team/junit5/issues/4663#issuecomment-2991083229.
